wxListCtrl fixes for wxQt

AliKet opened the pull request:
The following events are still not implemented by this PR:
```
wxEVT_LIST_BEGIN_DRAG
wxEVT_LIST_BEGIN_RDRAG
wxEVT_LIST_COL_BEGIN_DRAG
wxEVT_LIST_CACHE_HINT
```
other than that, this PR makes `[ListCtrlTestCase]` and `[VirtListCtrlTestCase]` tests pass for me.

Commit:
wxQt: Fix bug in wxListCtrl with checkboxes

Don't return the boolean value (m_checked) from wxQtListModel::data()
since doing so will actually return Qt::PartiallyChecked if m_checked
is true, and not Qt::Checked (enum value 2) which is not what we want.

Commit:
wxQt compatibility fix: wxListCtrl::GetColumnCount() always return 1 for wxLC_LIST

Commit:
wxQt: implement wxListCtrl::IsVisible()

Commit:
wxQt: fix wxListCtrl::GetCountPerPage() implementation

Use QTreeView's viewport height to get the right count per page
Also document that the control must contain at least one item in
order to return a valid value from the function.

Commit:
wxQt: Make wxListCtrl::SetColumnWidth() conformant to documentation

Also update the listctrl sample to show it's effect on the second column
wxLIST_AUTOSIZE (Ctrl+R) to resize the column to content
wxLIST_AUTOSIZE_USEHEADER (Ctrl+Shift+R) to resize it to header section

Commit:
wxQt: wxListCtrl with icons support get done

Commit:
wxQt: fix wxQtListModel::SetColumn() implementation

Using the wxListItem's mask and state mask, one can change only selected
attributes of a wxListCtrl item.

Commit:
wxQt: handle wxLIST_RECT_XXX code passed to wxListCtrl::GetSubItemRect()

This makes the test pass now as GetSubItemRect() can report subitem rectangles
correctly, labels or icons. icons support is added in this commit e6ad3e1340
(wxListCtrl with icons support get done)

Commit:
wxQt: fix crash when SetItem() is called for virtual wxListCtrl

Commit:
wxQt: remove redundant code which is also done in SetWindowStyleFlag()

Commit:
wxQt: fix creating the wxListCtrl with wxLC_EDIT_LABELS flag set

Also make EditLabel() implicitly selects and focuses the specified item
as the native control under wxMSW does. this actualy fixes two bugs in wxQt
1) EndEditLabel() will be able to find the item being edited.
2) The control display multiple in-place editors in different areas the
first time EditLabel() is called.

Commit:
wxQt: enable navigation between items using TAB key

Commit:
wxQt: Enhance EmitListEvent() to support more list-event types

We will need it in the upcoming commits

Commit:
wxQt: fix wxEVT_LIST_ITEM_XXX events generation

currentChanged() generates wxEVT_LIST_ITEM_FOCUSED event

itemActivated() generates wxEVT_LIST_ITEM_ACTIVATED event

selectionChanged() generates wxEVT_LIST_ITEM_SELECTED and wxEVT_LIST_ITEM_DESELECTED events

itemPressed() generates wxEVT_LIST_ITEM_RIGHT_CLICK and wxEVT_LIST_ITEM_MIDDLE_CLICK events

Commit:
wxQt: add wxQtHeaderView for wxEVT_LIST_COL_XXX events generation

Commit:
wxQt: added support for sort indicators

AliKet commented on src/qt/listctrl.cpp:
```diff
@@ -2108,6 +2113,36 @@ bool wxListCtrl::SortItems(wxListCtrlCompare fn, wxIntPtr data)
+        header->setSortIndicator(col, ascending ? Qt::DescendingOrder
+                                                : Qt::AscendingOrder);
+}
+
```
This should be
`        header->setSortIndicator(col, ascending ? Qt::AscendingOrder : Qt::DescendingOrder);`

But there seems to be a bug somewhere which is causing it to not work as expected!

vadz replied:
Do you mean that it works right now, with the inverted logic? If yes, then we must not be using the same meaning for the "ascending" order as Qt does and I guess we just need to add a comment documenting it here.

AliKet replied:
Yes, with the inverted logic everything is ok.

After building a simple Qt program (using **QtCreator**), I discovered that the sort indicator symbol is indeed reversed!

I'll just add a comment documenting it then.

vadz commented on src/qt/listctrl.cpp:
```diff
@@ -417,17 +442,37 @@ class wxQtListModel : public QAbstractTableModel
         RowItem &rowItem = m_rows[row];
         ColumnItem &columnItem = rowItem[col];
 
-        QVector<int> roles;
-
```
Not sure why was this moved above, there doesn't seem to be any reason for doing this?

vadz commented on src/qt/listctrl.cpp:
```diff
@@ -911,10 +949,20 @@ class wxQtVirtualListModel : public wxQtListModel
+
+    // Data type passed to EmitListEvent() as extra data which is essentially
+    // a pair of two ints with these meanings:
+    // first  = column or first (de)selected row
+    // second = column width or last (de)selected row
```
Would it be much more difficult to declare this as a struct with the named fields describing their meanings? I find the code using `pair` much more difficult to use than using named fields.

vadz commented on src/qt/listctrl.cpp:
```diff
@@ -980,9 +1039,146 @@ class wxQtListTreeWidget : public wxQtEventSignalHandler< QTreeView, wxListCtrl
+
+        if ( !itemPressed.isValid() )
+        {
+            // wx: deselect all if clicking on empty space
```
I don't think it's a requirement of wx API, so if Qt applications don't do this, e.g. if clicking on an empty space doesn't change the selection in standard KDE applications, I think it would make more sense not to do it in the wxQt ones neither (and updated/exclude the tests relying on this happening if we have any).

vadz commented on src/qt/listctrl.cpp:
```diff
@@ -980,9 +1039,146 @@ class wxQtListTreeWidget : public wxQtEventSignalHandler< QTreeView, wxListCtrl
+
+        wxQtListTreeWidget* const m_parent;
+
+        bool m_isDragging;
```
I'd initialize it here because there is just no reason not to:
```suggestion
        bool m_isDragging = false;
```
(and then we could remove its initialization from ctor member initializer list, of course).

vadz commented on src/qt/listctrl.cpp:
```diff
@@ -994,41 +1190,99 @@ wxQtListTreeWidget::wxQtListTreeWidget( wxWindow *parent, wxListCtrl *handler )
+                // one for each range which is more efficient (see issue #4541)
+                // data->first is the first row in the (de)selection
+                // data->second is the last row in the (de)selection
+                wxFAIL_MSG("No implemention yet");
```
```suggestion
                wxFAIL_MSG("No implementation yet");
```

But, more importantly, should we really have this assert at all here? This risks being annoying when using the control as a lot of user actions would result in an assert...

AliKet replied:
But no assertion is triggered yet and this is just a reminder for the maintainer if he tries to do a call like this:
```
    const ListEventData data { firstSel, lastSel };
    EmitListEvent(wxEVT_LIST_ITEM_SELECTED, QModelIndex(), &data);
```
which is supposed to send only one `wxEVT_LIST_ITEM_SELECTED` message and the handler can use something like: `event.GetRange();` to get the selected range generating the event (likewise for `wxEVT_LIST_ITEM_DESELECTED`)

but this is the subject of resolving this issue #4541

vadz commented on src/qt/listctrl.cpp:
```diff
@@ -1133,13 +1385,42 @@ int wxListCtrl::GetColumnWidth(int col) const
     {
-        m_qtTreeWidget->resizeColumnToContents(width);
-        return true;
+        // Always stretch the last section if _with_ is either
```
```suggestion
        // Always stretch the last section if _width_ is either
```

vadz commented on src/qt/listctrl.cpp:
```diff
@@ -1133,13 +1385,42 @@ int wxListCtrl::GetColumnWidth(int col) const
 
-    m_qtTreeWidget->setColumnWidth(col, width);
+    if ( width >= 0 )
+        m_qtTreeWidget->setColumnWidth(col, width);
```
Having braces around `else` but not `if` branch looks confusing, let's use them here too:
```suggestion
    {
        m_qtTreeWidget->setColumnWidth(col, width);
    }
```

vadz reviewed (commented):
Thanks for your work on this!

No real comments, I'm only not sure about leaving the "not implemented" assert as I think this is more annoying than useful.

Commit:
Apply review suggestions

Commit:
fix compilation with msvc2015

vadz commented:
Thanks, should this be merged now or do you plan any more changes here?

Also, I'm a bit worried by the last commit: does MSVS 2015 really have a problem with member initialization? This would be sad, as I really like using it...

AliKet commented:
> Thanks, should this be merged now or do you plan any more changes here?
> 

Just a commit for `wxEVT_LIST_KEY_DOWN` event generation.

> Also, I'm a bit worried by the last commit: does MSVS 2015 really have a problem with member initialization? This would be sad, as I really like using it...

Sorry for the unclear commit message!

What doesn't really work with MSVS 2015 (later versions do not have this bug) is when you try to brace initialize a `class/struct` which has in-class member initialization. although the two features are supported by that version of compiler see:

https://learn.microsoft.com/en-us/cpp/cpp/initializing-classes-and-structs-without-constructors-cpp?redirectedfrom=MSDN&view=msvc-140

https://learn.microsoft.com/en-us/cpp/cpp/class-member-overview?view=msvc-140

However, I couldn't find a formal bug report for this specific issue!

Commit:
wxQt: fix crash when asterisk key (Qt::Key_Asterisk) is pressed

The crash happens inside QTreeView::keyPressEvent(QKeyEvent*) when handling
Qt::Key_Asterisk key press. The reason is the non conformant rowCount() and
columnCount() implementations which according to Qt documentation should
return 0 when the parent is valid.

Commit:
wxQt: added wxEVT_LIST_KEY_DOWN event generation to wxListCtrl

AliKet commented:
@vadz  I think i've done with this for now, i.e. the tests and the listctrl sample work as expected. the missing functionalities will be added later, support for drag and drop for example.

Thanks for the review.

vadz commented:
Merged now, thanks again!
